sys/net/application_layer/coap/gcoap.c

```diff
@@ -587,12 +587,13 @@ kernel_pid_t gcoap_init(void)
     _pid = thread_create(_msg_stack, sizeof(_msg_stack), THREAD_PRIORITY_MAIN - 1,
                             THREAD_CREATE_STACKTEST, _event_loop, NULL, "coap");
 
+    mutex_init(&_coap_state.lock);
     /* Blank lists so we know if an entry is available. */
     memset(&_coap_state.open_reqs[0], 0, sizeof(_coap_state.open_reqs));
     memset(&_coap_state.observers[0], 0, sizeof(_coap_state.observers));
     memset(&_coap_state.observe_memos[0], 0, sizeof(_coap_state.observe_memos));
     /* randomize initial value */
-    _coap_state.last_message_id = random_uint32() & 0xFFFF;
+    atomic_init(&_coap_state.next_message_id, (uint16_t)random_uint32());
```

Comment:
from what I read it might be safer to use `atomic_store` here. A quick grep through the RIOT source code also shows, that it is favored (and used only) over `atomic_init`, see [1].

[1]: https://github.com/RIOT-OS/RIOT/blob/master/tests/emb6/ping.c#L119